### ISSUE TITLE
Doubt on the dependency on pkgs(indirectly on systems) where it may (possibly) not be required

### ISSUE THREAD

**Issue by paulvictor** [terranix/terranix#108]
Hello,

Want to reach out and provide a heartfelt thanks for this wonderful infrastructure.

Something I felt when I was browsing through the existing modules.

The core(defined in `core/default.nix`) depends on pkgs(and on system).
When I look at usages though, there are a couple of usages of lib(which can be replaced with `inputs.nixpkgs.lib`) and a reference to pkgs.writeFile [here](https://github.com/terranix/terranix/blob/main/core/helpers.nix#L19) which can be replaced with builtins.toFile.
While I am not a nix expert(and I stand to be corrected), want to know the rationale behind this. The reason I am asking is that looks like we are generating a json file at the end of the day and adding a ${system} inbetween looks weird.
Please let me know if I have missed anything.
PS: Currently I am using the 2.7.0 version of this without flake-parts and have to switch to the next version yet.

Thanks
Paul

**Comment by terlar** [comment]
This is very valid and I am pretty sure your suggestions can be applied. I have another module system based generation of JSON files (generating GitLab/GitHub pipelines) and did what you described here and it worked. My motivation was to not have to pull nixpkgs during CI. Instead I was using flake-parts and nixpkgs-lib which is a subset and much quicker to fetch.

So doing this change + switching to nixpkgs-lib could greatly speed up the generation.

**Comment by adrian-gierakowski** [comment]
> pkgs.writeFile [here](https://github.com/terranix/terranix/blob/main/core/helpers.nix#L19) which can be replaced with builtins.toFile

String passed to toFile cannot contain references to outputs of derivations, while writeText can

**Comment by paulvictor** [comment]
Agreed @adrian-gierakowski , I didnt know about this(I always learn something new about nix). But just curious, can that be solved by passing pkgs as a separate `module._args`(by consumers of this module system) and using it in the modules which may need that facility while keeping the outer api free of pkgs/system ? Somehow the `system` reference looks very starkingly unnecessary. Now when I have to build a flake output, I have to specify a value for system(say x86_64-linux)

**Comment by sshine** [comment]
I see that nobody has acted on this issue, although it appears to be low-hanging fruit. I've tagged it "good first issue" and will get back to it myself if nobody chimes in.
